main.go

```diff
@@ -55,8 +73,105 @@ func echo(w http.ResponseWriter, r *http.Request) {
 			logged = true
 		}
 
-		if strings.Contains(string(message), "node-ping") {
+		if strings.Contains(string(message), "pending") {
+
+			var msg map[string]interface{}
+			if err := json.Unmarshal(message, &msg); err != nil {
+				log.Println(err)
+			}
+
+			// fmt.Println("lol : ", (msg["emit"]).([]interface{})[1])
+
+		} else if strings.Contains(string(message), "REORGS DETECTED") {
+
+			var msg map[string]interface{}
+			if err := json.Unmarshal(message, &msg); err != nil {
+				log.Println(err)
+			}
+
+			block_number := int((msg["emit"]).([]interface{})[1].(map[string]interface{})["block"].(map[string]interface{})["number"].(float64))
```

Comment:
This is quite messy and super hard to understand. I am assuming that the message was something like this:
```
{
   "emit": [
      "..",
      {
          "block": {} // block
      }
   ]
}
```
This is a complex message but you should have at least documented the format in the code.
How would I do this less messy?
Have some wrapper function that decodes the 1 object in the emit message since that is where the data is always:
```
type Msg struct {
   Data map[string]json.RawMessage
}

func decodeMsg(message []byte) (*Msg, error) {
   var decodeRaw struct {
      Emit []json.RawMesssge
   }
   if err :=  json.Unmarshal(message); err != nil {
      return nil, err
   }
   // now decode the map at pos 1
   var data map[string]json.RawMessage
   if err := json.Unmarshal(&data, decodeRaw.Emit[1]); err != nil {
      return nil, err
   }
   msg := &Msg{
      Data: data,
   }
   return msg, nil
}
```
Now, `Msg` has a specific function to decode a key in that data map to a given struct.
```
func (m *Msg) Decode(key string, obj interface{}) error {
    return json.Unmarshal(obj, m.Data[key])
}
```
Then, you have to create a struct object to represent the block that can be unmarshalled to that json file. Hopefully, there is already one in web3.Block library [here](https://github.com/umbracle/go-web3).

Now, you can combine everything and do:
```
block := web3.Block{}
msg, err := decodeMsg(message)

if err := msg.Decode("block", &block); err != nil {
   panic(err)
}

state.WriteBlock(block)
```

Reply:
I made the code compact with the help of your idea and without the need of the JSON file.